### PR TITLE
Node: Use objects over multiple parameters

### PR DESCRIPTION
**Pull request by landrito**
Resolves [gax-nodejs/48](https://github.com/googleapis/gax-nodejs/issues/48)

### CODE REVIEW

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 1]
```diff
@@ -240,14 +241,25 @@ private String fieldTypeCardinalityComment(Field field) {
...
+  /**
    * Returns a JSDoc comment string for the field as a parameter to a function.
    */
-  private String fieldParamComment(Field field, String paramComment, boolean isOptional) {
+  private String fieldParamComment(Field field, String paramComment, ParamType paramType) {
     String commentType = fieldTypeCardinalityComment(field);
-    String fieldName = wrapIfKeywordOrBuiltIn(lowerUnderscoreToLowerCamel(field.getSimpleName()));
-    if (isOptional) {
+    String fieldName = lowerUnderscoreToLowerCamel(field.getSimpleName());
+    if (paramType == ParamType.OPTIONAL) {
```
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Oh, this is not my intention here -- I think all of the parameters are under `request.`. `options` will contain call options and that's it -- this is more usable.

Comment (landrito) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Ah I'll see.

Comment (landrito) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Fixed.

**Comment by landrito** [comment]
@jmuk This is ready for another pass. PTAL

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 2]
```diff
@@ -379,21 +380,19 @@ private String returnTypeComment(Method method, MethodConfig config) {
    */
   public List<String> methodComments(Interface service, Method msg) {
     MethodConfig config = getApiConfig().getInterfaceConfig(service).getMethodConfig(msg);
-
     // Generate parameter types
     StringBuilder paramTypesBuilder = new StringBuilder();
+    if (config.getRequiredFields().iterator().hasNext()) {
```
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Hmm, I think this is not right, considering some method whose request does not have any required fields but many optional fields. Possibly `config.getRequiredFields().iterator().hasNext() || config.getOptionalFields().iterator().hasNext())` might be a better condition (a bit verbose though).

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java [hunk 3]
```diff
@@ -14,6 +14,7 @@
  */
 package com.google.api.codegen.nodejs;
 
+import autovalue.shaded.com.google.common.common.collect.Lists;
```
Comment (jmuk) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: which part of the code requires this?

File: src/main/resources/com/google/api/codegen/nodejs/main.snip [hunk 4]
```diff
@@ -350,64 +350,62 @@
...
-        @end
-      @end
-      return this._{@methodName(method)}(req, options, callback);
+      return this._{@methodName(method)}({@requestObject(apiMethodView)}, options, callback);
     };
   @end
 @end
 
+@private requestObject(apiMethodView)
+    @if apiMethodView.requiredRequestObjectParams.isEmpty
```
Comment (jmuk) [inline src/main/resources/com/google/api/codegen/nodejs/main.snip]: nit: It seems this file usually has 2-space indents. Be consistent.

File: src/main/resources/com/google/api/codegen/nodejs/main.snip [hunk 5]
```diff
@@ -350,64 +350,62 @@
...
+@end
+
+@private methodRequestParameters(apiMethodView)
+    @if apiMethodView.requiredRequestObjectParams.isEmpty
+    @else
+        request, {@""}
+    @end
+@end
+
 @private fieldParameterName(field)
```
Comment (jmuk) [inline src/main/resources/com/google/api/codegen/nodejs/main.snip]: Remove this -- all usage of `fieldParameterName` seems to be gone.

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 3 above)
Comment (landrito) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Woops. removed.

File: src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java (hunk 2 above)
Comment (landrito) [inline src/main/java/com/google/api/codegen/nodejs/NodeJSGapicContext.java]: Ah Good catch. Changed.

File: src/main/resources/com/google/api/codegen/nodejs/main.snip (hunk 4 above)
Comment (landrito) [inline src/main/resources/com/google/api/codegen/nodejs/main.snip]: Woops! Changed.

File: src/main/resources/com/google/api/codegen/nodejs/main.snip (hunk 5 above)
Comment (landrito) [inline src/main/resources/com/google/api/codegen/nodejs/main.snip]: Removed.

**Comment by landrito** [comment]
Removed wrapIfKeywordOrBuiltIn and KEYWORD_BUILT_IN_SET.
Ready for another pass!

**Pull request merged**